Add a readme.md to the models dir. Rename the transformers corpora. (#76)
groq/mlagility: models/readme.md

# MLAgility Benchmark

This directory contains the MLAgility benchmark, which is a large collection of models that can be evaluated using the [`benchit` CLI tool](https://github.com/groq/mlagility/blob/main/docs/benchit_user_guide.md).

## Table of Contents

- [Benchmark Organization](#benchmark-organization)
- [Running the Benchmark](#running-the-benchmark)
- [Model Template](#model-template)

## Benchmark Organization

> The MLAgility benchmark is made up of several corpora of models (_corpora_ is the plural of _corpus_... we had to look it up too). Each corpus is named after the online repository that the models were sourced from. Each corpus gets its own subdirectory in the `models` directory. 

Reviewer: Do we want to create a distinction between scripts and models (each script may instantiate multiple models)?

Author: Yes, added!

- `diffusers`: models from the [Huggingface `diffusers` library](https://huggingface.co/docs/diffusers/index), including the models that make up Stable Diffusion.
- `graph_convolutions`: Graph Neural Network (GNN) models from a variety of publications. See the docstring on each .py file for the source.
- `popular_on_huggingface`: hundreds of the most-downloaded models from the [Huggingface models repository](https://huggingface.co/models).
- `selftest`: a small corpus with small models that can be used for testing out the MLAgility tools.
- `torch_hub`: a variety of models, including many image classification models, from the [Torch Hub repository](https://github.com/pytorch/hub).
- `torchvision`: image recognition models from the [`torchvision` library](https://pytorch.org/vision/stable/index.html).
  - _Note_: the `torchvision` library also includes many image classification models, but we excluded them to avoid overlap between our `torch_hub` and `torchvision` corpora.
- `transformers_keras`: Transformer models from the [Huggingface `transformers` library](https://huggingface.co/docs/transformers/index), expressed in the Keras framework.

> - `transformers_pytorch`: Transformer models from the [Huggingface `transformers` library](https://huggingface.co/docs/transformers/index), expressed in the PyTorch framework.

Reviewer: There is no one-to-one match between the `transformers_keras` and the `transformers_pytorch` corpora. However, overlaps do exist.

Author: Added!

## Running the Benchmark

### Prerequisites

Before running the benchmark we suggest you familiarize yourself with the [`benchit` CLI tool](https://github.com/groq/mlagility/blob/main/docs/benchit_user_guide.md) and some of the [`benchit` CLI examples](https://github.com/groq/mlagility/tree/main/examples/cli).

You must also run the following command to install all of the models' dependencies into your Python environment.

`pip install -r models/requirements.txt`

### Benchmarking Commands

Once you have fulfilled the prerequisites, you can evaluate one model from the benchmark with a command like this:

```
cd MLAGILITY_ROOT/models # MLAGILITY_ROOT is where you cloned mlagility
benchit benchmark linear.py --search-dir selftest
```

You can also evaluate an entire corpus with a command like this:
```
cd MLAGILITY_ROOT/models # MLAGILITY_ROOT is where you cloned mlagility
benchit benchmark --all --search-dir selftest
```

You can aggregate all of the benchmarking results from your `mlagility cache` into a CSV file with:

```
benchit report
```

> If you want to only report on a subset of models, we recommend benchmarking those models into a specific cache directory:

Reviewer: ```suggestion
If you want to only report on a subset of models, we recommend saving the benchmarking results into a specific cache directory:
```
Suggestion for clarity

Author: nice

```

> # Benchmark the models into a specific cache directory

Reviewer: ```suggestion
# Save benchmarked models into a specific cache directory
```

Author: Made an edit like this

benchit benchmark --all --search-dir selftest --cache-dir selftest_results

# Report the results from the `selftest_results` cache
benchit report --cache-dir selftest_results
```

## Model Template

Each model in the MLAgility benchmark follows a template to keep things consistent. This template is meant to be as simple as possible. The models themselves should also be instantiated and called in a completely vendor-neutral manner.

### The model.py File

> Each model in the MLAgility benchmark is hosted in a `model.py` file. This file is a Python script that must meet the following requirements:

Reviewer: Is the idea to only allow scripts in mlagility if that script only instantiates a single model?

Author: Nope, reworked the definition

Each `model.py` can optionally include a set of [labels](#labels) and [parameters](#parameters).

### Labels

MLAgility uses labels to help organize the results data. Labels must be in the first line of the Python file and start with `# labels: `

Each label must have the format `key::value1,value2,...`

Example:

```
# labels: author::google test_group::daily,monthly
```
     
Labels are saved in your cache directory and can later be retrieved using the function `mlagility.common.labels.load_from_cache()`, which receives the `cache_dir` and `build_name` as inputs and returns the labels as a dictionary. 

### Parameters

MLAgility supports parameterizing models so that you can sweep over interesting properties such as batch size, sequence length, image size, etc. The `mlagility.parser.parse()` method parses a set of standardized parameters that we have defined and provides them to the model as it is instantiated.

For example, this code would retrieve user-defined `batch_size` and `max_seq_length` (maximum sequence length) parameters.

```
from mlagility.parser import parse
parse(["batch_size", "max_seq_length"])
```

> You can pass parameters into a benchmarking run with the `--script-args` argument to `benchit`. For example, the command:

Reviewer: Might be useful to explain that those CLI arguments may also be used outside of benchit

Example `bert.py --batch-size 8`

Author: Added!

```

> benchit bert.py --search-dir transformers_pytorch --script-args="--batch-size=8,--max-seq-length=128"

Reviewer: ```suggestion
benchit bert.py --search-dir transformers_pytorch --script-args="--batch-size 8 --max-seq-length 128"
```

Author: good catch thanks

```

would set `batch_size=8` and `max_seq_length=128` for that benchmarking run.

The standardized set of arguments is:

- General args
    - "batch_size": Arg("batch_size", default=1, type=int),
    - "max_seq_length": Arg("max_seq_length", default=128, type=int),
    - "max_audio_seq_length": Arg("max_audio_seq_length", default=25600, type=int),
    - "height": Arg("height", default=224, type=int),
    - "num_channels": Arg("num_channels", default=3, type=int),
    - "width": Arg("width", default=224, type=int),
- Args for Graph Neural Networks
    - "k": Arg("k", default=8, type=int),
    - "alpha": Arg("alpha", default=2.2, type=float),
    - "out_channels": Arg("out_channels", default=16, type=int),
    - "num_layers": Arg("num_layers", default=8, type=int),

>     - "in_channels": Arg("in_channels", default=1433, type=int),- 

Reviewer: We might want to review those eventually, as a few of those GNN parameters are only used in 1-2 models and are not very descriptive.

Author: Yeah they should probably have comments or something

Author: I added definitions for the general args. Made an issue for @ramkrishna2910 to add definitions to the GNN args https://github.com/groq/mlagility/issues/78
